### ISSUE TITLE
Connection to Cockroach not re-established on time in case of a CRDB node failure

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#17680]
### What version of Materialize are you using?

v0.44.0-dev (2bc1fe19d)

### How did you install Materialize?

Docker image

### What is the issue?

If one node in a CRDB cluster goes down , then Mz does not quickly re-establish connection to the other available nodes.

The following variants have been observed:

1. If the individual CRDB nodes are provided as individual host names in the crdb connection string and one of them becomes unresolvable , Mz will continue to attempt to resolve it , rather than connect to another of the hostnames (This was only seen in testing, as in production/staging our crdb connection string contains only one hostname)

2. If all crdb nodes are behind the same hostname (which is the case with Cockroach cloud), and resolve to multiple IPs via DNS, killing one of the crdb nodes causes Mz to report the following:
```
environmentd: 2023-02-15T09:53:11.128497Z  INFO deadpool.postgres: Connection could not be recycled: Connection closed    
```
and proper operation is only restored after several minutes.

It may be the case that tokio_postgres and the deadpool and not perfectly set up for load balancing. In addition to the symptoms of the ticket, another piece of evidence to that effect is the way multiple hostnames are specified in the connection string: you are supposed list of hostnames *first* , followed by a list of TCP ports *after*. This does not follow https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING and tells me that this functionality was not well thought out from the start and was bolted on after the fact.

To reproduce:

1. Clone `https://github.com/philip-stoev/materialize/pull/new/crdb-rolling-restarts`
2. Do:
```
cd test/crdb-rolling-restart
./mzcompose down -v ;
./mzcompose run default ;
docker kill crdb-rolling-restart-cockroach2-1 ;
echo -e "\\\timing on \n DROP TABLE IF EXISTS t1; CREATE TABLE t1 (f1 INTEGER); " | psql -p  `docker port crdb-rolling-restart-materialized-1 | grep '0.0' | grep 6875 | cut -d':' -f2` -h 127.0.0.1 -U materialize
```

What this does is start a 3-node CRDB cluster and then Mz on top of it. It will then kill one of the nodes of that cluster. If the node that was killed is the one Mz connected to (you may need to try multiple times), the following will be observed:

A. The DDLs issued against Mz will be very slow or hanging:

```
Timing is on.
DROP TABLE
Time: 129753.464 ms (02:09.753)
CREATE TABLE
Time: 437.412 ms
```
B. The log of the materialized container will be full with the `connection could not be recycled` messages.

The CRDB cluster may theoretically be degraded by the killing of one node, but it appears to still be operational, at least when it comes to creating new tables directly on it.

**Comment by aljoscha** [comment]
@danhhz Not 100% obvious from the issue description, but I think this is a persist thing because the Stash doesn't use `deadpool-postgres`.

**Comment by pH14** [comment]
Thanks for the detailed write up and repro steps here Philip! I hooked up tracing to the repro and can confirm the slowdown is coming somewhere within storage. I'll keep digging tomorrow to see where all the time is going

**Comment by benesch** [comment]
Porting over a comment from Slack:

> The test that would most directly simulate production is to run an HA proxy with at least two nodes and a Cockroach cluster with at least three nodes, have the HA proxy do health checks for all Cockroach nodes, set up one hostname to DNS resolve to both HA proxy nodes, point Materialize at that hostname, and then see what happens if you restart one of the HA proxy nodes and what happens if you restart one of the Cockroach nodes. Materialize really should survive both of these scenarios with minimal downtime.

We should be careful here: I think the _multiple hostname_ scenario is not immediately relevant to Materialize, because our production/staging environments have only a single hostname. 

@philip-stoev I'm not entirely sure what happens under the hood with your test. Docker Compose has this to say:

![image](https://user-images.githubusercontent.com/882976/219264961-b25531df-07f7-4bca-a94c-161b58a21a7c.png)

So it is possible that we are observing Docker badness here, and not Cockroach badness.

Before we go further in this investigation, I'd suggest we set up an actual HAProxy service in the test, to better simulate how this works in production.

**Comment by philip-stoev** [comment]
@benesch I will add HAProxy later but unfortunately docker, for once, is doing the right thing in this case. Consider the following script:

```
./mzcompose down -v
./mzcompose run default
docker kill crdb-rolling-restart-cockroach2-1
docker exec -it crdb-rolling-restart-materialized-1 bash -c 'PGCLIENTENCODING="utf-8" echo -e "\\\timing on \n CREATE TABLE table_in_crdb (f1 INTEGER); SELECT  COUNT(*) FROM consensus.consensus; INSERT INTO consensus.consensus VALUES (1,1,1::char::bytes);" | psql -h cockroach -p 26257 -U root defaultdb'
echo -e "\\\timing on \n DROP TABLE IF EXISTS table_in_mz; CREATE TABLE table_in_mz (f1 INTEGER); " | psql -p  `docker port crdb-rolling-restart-materialized-1 | grep '0.0' | grep 6875 | cut -d':' -f2` -h 127.0.0.1 -U materialize
```

In other words, before we attempt to issue DDL on Materialize, we connect to CRDB **from the Mz container using the same hostname Mz would use** and confirm that CRDB is fully operational -- you can create tables, read and write to the consensus table.

The outcome of this script is like this:

```
+ docker exec -it crdb-rolling-restart-materialized-1 bash -c 'PGCLIENTENCODING="utf-8" echo -e "\\\timing on \n CREATE TABLE table_in_crdb (f1 INTEGER); SELECT  COUNT(*) FROM consensus.consensus; INSERT INTO consensus.consensus VALUES (1,1,1::char::bytes);" | psql -h cockroach -p 26257 -U root defaultdb'
Timing is on.
CREATE TABLE
Time: 25.028 ms <- FAST
 count 
-------
   385
(1 row)

Time: 24.632 ms <- FAST
INSERT 0 1
Time: 9.455 ms
+ echo -e '\\timing on \n DROP TABLE IF EXISTS table_in_mz; CREATE TABLE table_in_mz (f1 INTEGER); '
++ docker port crdb-rolling-restart-materialized-1
++ grep 0.0
++ grep 6875
++ cut -d: -f2
+ psql -p 56080 -h 127.0.0.1 -U materialize
Timing is on.
^CCancel request sent
DROP TABLE
Time: 129798.581 ms (02:09.799) <- SLOW
CREATE TABLE
Time: 407.725 ms
```

In other words, the SQL we issued directly to CRDB all succeeded immediately while Mz still hung for 2+ minutes.

Furthermore, while Mz was hung , in a separate console we can connect to CRDB again and confirm that it is still operational:

```
docker exec -it crdb-rolling-restart-materialized-1 bash -c 'PGCLIENTENCODING="utf-8" echo -e "\\\timing on \n CREATE TABLE table_in_crdb (f1 INTEGER); SELECT  COUNT(*) FROM consensus.consensus; INSERT INTO consensus.consensus VALUES (1,1,1::char::bytes);" | psql -h cockroach -p 26257 -U root defaultdb'
Timing is on.
ERROR:  relation "defaultdb.public.table_in_crdb" already exists
Time: 5.561 ms
 count 
-------
   386
(1 row)

Time: 17.147 ms
ERROR:  duplicate key value violates unique constraint "consensus_pkey"
DETAIL:  Key (shard,sequence_number)=('1',1) already exists.
Time: 2.237 ms
```

So, if the psql client can connect to CRDB from within the Mz container and using the same hostname Mz would use to connect, and is able to perform DML , then Mz should also be able to do the same by reconnecting. This way it would have been able to continue processing and complete the DDL without hanging for the 2+ minutes.

**Comment by benesch** [comment]
Ok, cool, that's great! Thanks @philip-stoev! Sounds like it is worth going ahead with the investigation with the current test harness. Hopefully it's an easy fix. 🤞🏽

**Comment by pH14** [comment]
Hmm I'm running on an M1 so maybe the characteristics are different, but I'm seeing much shorter reconnections from Mz:

```
echo -e "\\\timing on \n DROP TABLE IF EXISTS t1; CREATE TABLE t1 (f1 INTEGER); " | psql -p  `docker port crdb-rolling-restart-materialized-1 | grep '0.0' | grep 6875 | cut -d':' -f2` -h 127.0.0.1 -U materialize
Timing is on.
DROP TABLE
Time: 9911.950 ms (00:09.912)
CREATE TABLE
Time: 210.783 ms
```

^^ This is much more typical for my test runs. While a 10s DDL isn't great, that's definitely a lot better than 120s.

Additionally, contrary to my other statement, I've now seen some slow traces where the slowness comes from stash queries:

<img width="1701" alt="Screenshot 2023-02-15 at 10 29 13 PM" src="https://user-images.githubusercontent.com/785446/219400181-277aba87-b42a-4509-9bd0-03a953b2af85.png">

In addition to slow persist queries:

<img width="1695" alt="Screenshot 2023-02-16 at 9 59 50 AM" src="https://user-images.githubusercontent.com/785446/219400557-44fd6228-2d04-462b-b1ad-b7b00d375dc5.png">

The long spans in both persist and stash are both entirely within `tokio-postgres` at this point, so I think the ~10s reconnection time is below a level that we obviously control. I've additionally tried setting a very low `connect_timeout` for all of our clients, to no effect (which makes sense, since the issue seems to be that an existing connection gets mired connecting to a now-terminated host rather than failure to create a new connection).

I'm wondering if part of the issue here is lack of client-side timeouts, that's the next item to test out

**Comment by pH14** [comment]
OK I tried adding in very short client-side timeouts to both stash and persist, and the downtime is about the same as before. Rather than 1 long `tokio-postgres` call to CRDB, we see a beautiful waterfall of timed out attempts:

<img width="1703" alt="Screenshot 2023-02-16 at 11 13 10 AM" src="https://user-images.githubusercontent.com/785446/219424756-bbf7ac68-8156-4a36-87cf-a1b2aa2ff235.png">

I hopped on port 8080 of one of the CRDB nodes, and saw a corresponding drop in activity that spans a similar time interval. CRDB reports its numbers every 10s, and we see thing healthy at the 30s and 50s mark, but down at the 40s mark. The trace started at 35s past, so we know its downtime spanned at least 5s.

<img width="846" alt="Screenshot 2023-02-16 at 11 28 49 AM" src="https://user-images.githubusercontent.com/785446/219428560-070a97d9-415d-40d8-8cd2-88ac0e2cda99.png">

At least locally, it seems like the Mz downtime maps on quite closely to the CRBD downtime. I'm not sure why I'm only seeing ~10s downtime each time rather than the ~120s that Philip is seeing -- @philip-stoev are you running on a scratch machine or locally on a Linux box?

**Comment by philip-stoev** [comment]
@pH14  i think the crdb cluster has no downtime. If you connect to a working node you can see that it is operational throughout. I think your trace indicates that mz attempts to connect over and over to a host that is already down rather than properly trying another host and be successfull there.

I used a Linux box I can give you access to if this will help with debugging. I have not tried scratch

**Comment by philip-stoev** [comment]
@pH14 , yes you are right there a 9-second delay as CRDB adjusts . So indeed it is not instantanous, as I previously reported.

Anyway, the issue is reproducible on a scratch ci machine , you just need to run it in a loop until it hits:

```
cat ./script.sh 
./mzcompose down -v
./mzcompose run default
docker kill crdb-rolling-restart-cockroach2-1
echo -e "\\\timing on \n DROP TABLE IF EXISTS table_in_mz; CREATE TABLE table_in_mz (f1 INTEGER); " | psql -p  `docker port crdb-rolling-restart-materialized-1 | grep '0.0' | grep 6875 | cut -d':' -f2` -h 127.0.0.1 -U materialize
```
```
while true ; do  bash -x -c "./script.sh 2>&1  | grep Time" ; done
```
Leave it to run 10 or 20 cycles. The output will have lines such as:

```
Time: 68233.403 ms (01:08.233)
```

The **GOOD NEWS** is that your branch resolves the issue in question. Running the same script against it results in a maxium query delay of :

```
Time: 9324.039 ms (00:09.324)
```

Which is totally acceptable. If you like, you can prepare a PR with your fix without the test or any test-related python changes and then I will follow up with a separate PR with the test and have it reviewed independently.

**Comment by pH14** [comment]
Perfect! Thanks so much for verifying that Philip 🙇 

@mjibson while testing, I was able to find slow reconnections during CRDB recovery from stash transactions. I added [a crude client-side timeout](https://github.com/MaterializeInc/materialize/compare/main...pH14:materialize:crdb-rolling-restart-investigation#diff-0b48b05a131e1b397337dfa18afeee83665820f0f3e541fb1b9d399313b5d076) for the PR that Philip tested above, would you like me to include something similar in my PR? Or would you like to do that work separately? Offhand I'm not sure how to tune a timeout threshold for stash, because it seems like the expected amount of work a transaction must do is fairly variable

**Comment by mjibson** [comment]
I would like to do the stash work separately. It doesn't know what to do with indeterminate errors, so adding a tokio timeout would require more thinking to get right.

**Comment by benesch** [comment]
* I'm confused about what's going on in the case of the multi-minute hang. What layer is getting hung? Is the TCP connection stalling out? If so, would the situation be rectified if we instead enabled TCP keepalives with a short idle time (5s? 10s?)?
* I'm not convinced that we've gotten to the bottom of what's going on in staging/production. @umanwizard observed the hang while Cockroach was undergoing a _graceful_ restart, in which nodes should be neatly drained: https://www.cockroachlabs.com/docs/stable/node-shutdown.html#node-shutdown-sequence. Cockroach should be very deliberately closing connections during a graceful restart.

**Comment by philip-stoev** [comment]
@benesch I changed my script to do graceful restart via `cockroach node drain` (I also tried sending SIGTERM to the cockroach process, which is also a documented approach). In both cases I did not observe any stalls on the Mz side.

So, to summarize, the only way I was able to observe stall times consistent with the ones observed by @umanwizard was by hard-killing a CRDB node. There are two possible explanations:
- that the test does not faithfully reproduce the CRDB upgrade procedure that Cockroach are implementing on their cloud, and therefore my stall and @umanwizard's stalls are of a different nature. I could add a load-balancer in front, but this may or may not increase the faithfulness of the test 
- that Cockroach did not follow the published procedure when they performed the rolling upgrade. Certainly, the logs reported by @umanwizard did contain "ERROR: server is shutting down", which indicates orderly draining, but the stalls he reported are consistent with my observations about an unorderly shutdown.


One thing we could do is spin up 2 regions, one with and one without a patch. Direct a steady steam of DDLs against each and ask Cockroach to perform a rolling restart on the cluster that backs our staging environment.

**Comment by benesch** [comment]
Right, I think we need to throw a load balancer in the mix in order to more faithfully reproduce the production environment. I'm finding it hard to reason about how the failures you're observing translate to failures in production. Certainly the stall you found is a real bug, and we should fix it, but it's very hard to know whether it's at all related to what @umanwizard was observing in production, or a totally unrelated stall.

> One thing we could do is spin up 2 regions, one with and one without a patch. Direct a steady steam of DDLs against each and ask Cockroach to perform a rolling restart on the cluster that backs our staging environment.

I quite like the sound of this. In fact, I think I can give you an even better simulation:

  * Create a new CockroachDB dedicated cluster in AWS/us-east-1 in our Cockroach Cloud account. Call it "stall-test" or something like that. Give it three nodes, but use the smallest possible node size to save money.
  * Point a local Materialize instance at the dedicated cluster.
  * Trigger a cluster restart by scaling up the node sizes by one tick. Then trigger the next cluster restart by scaling _down_ the node sizes by one tick. Etc. A bit unfortunate to have to change the size to trigger the rolling restart, but I'm not aware of another way to trigger a rolling restart of _all_ nodes in a Cockroach cloud cluster.
  * While the cluster is restarting, monitor for stalls like @umanwizard observed in staging/production.

Using a separate CockroachDB cluster will avoid impacting other staging users with the scale up and scale down, but using CockroachDB Cloud should perfectly reproduce our staging/production environment. That way we won't have questions about whether the local HA proxy configuration matches the AWS LB configuration that Cockroach Cloud uses—we'll know we're getting the real deal!

**Comment by philip-stoev** [comment]
@benesch ok so I created a dedicated CockroachDB cluster and scaled it up and down:
- scaling up seems to work without hangs
- without @pH14's patch, the first scale-down had a 2+ min hang right off the gate
- with his patch, I was able to scale the cluster down from 7 to 3 nodes twice, with the longest hang being 30 seconds, while most hangs were less than 5 seconds each.

So I think his fix will also benefit the actual production setup . I will be tearing down the test CRDB cluster at the end of the day, so if you have any requests for additional experiments, please let me know today.

**Comment by benesch** [comment]
Oh, excellent, I’m glad it reproduced! Could you actually leave the cluster up and send me instructions for running the test?

@ph14’s change as it exists now is a little scary to merge, as fixed operation timeouts can lead to the system getting stuck in a retry loop forever. I want to try to understand the failure better, to see if there are other types of timeouts we could employ instead.

**Comment by philip-stoev** [comment]
To reproduce:

1. (Optionally) Use the `crdb-rolling-restart-investigation` branch from @pH14's Mz fork
2. Scale the "stall-test" cockroachDB to 7 nodes and wait for it to reach steady state
3. Initialize it:
```
drop schema adapter cascade ; drop schema consensus cascade ; drop schema storage cascade;
create schema adapter; create schema consensus ; create schema storage;
SET CLUSTER SETTING sql.stats.forecasts.enabled = false;
```
The SQL credentials are in the "Cockroach Cloud Stall Test Account" 1Password secret
5. Use the following mzcompose.py:
```
rom materialize.mzcompose import Composition
from materialize.mzcompose.services import Cockroach, Materialized, Testdrive, Service

CRDB_URL = 'cockroach:<PASSWORD>@<HOST>:26257'

SERVICES = [
    Materialized(
        options=[
            f"--adapter-stash-url=postgres://{CRDB_URL}?options=--search_path=adapter",
            f"--storage-stash-url=postgres://{CRDB_URL}?options=--search_path=storage",
            f"--persist-consensus-url=postgres://{CRDB_URL}?options=--search_path=consensus",
        ],
    ),
]


def workflow_default(c: Composition) -> None:
    c.up("materialized")
```
7. Create the following script:
```
cat ./script.sh 
echo -e "\\\timing on \n DROP TABLE IF EXISTS table_in_mz; CREATE TABLE table_in_mz (f1 INTEGER); " | psql -p  `docker port crdb-cloud-materialized-1 | grep '0.0' | grep 6875 | cut -d':' -f2` -h 127.0.0.1 -U materialize
```
8. and run it in a loop:
```
while true ; do  bash -x -c "./script.sh " ; done
```
9. Scale the CRDB cluster down to 3 nodes. You will observe that the timings of the DDLs as reported by the script above are perturbed as the cluster is scaled down.

**Comment by pH14** [comment]
@philip-stoev I've pushed a new branch `pH14:persist-crdb-tcp-keepalives` that relies on TCP keepalives to determine when to reconnect, would you be able to give that branch a spin?

There's some subtleties with using keepalives alone (https://blog.cloudflare.com/when-tcp-sockets-refuse-to-die/ has some great information) and it'd be better to also set `TCP_USER_TIMEOUT` on the postgres/CRDB sockets, but it looks like `tokio-postgres` doesn't support that right now. Might be interesting to add that feature to our branch and see if we can do everything we need in terms of TCP settings

**Comment by philip-stoev** [comment]
@pH14 first thing tomorrow.

**Comment by benesch** [comment]
Also, @philip-stoev, I suspect it will be more representative of what happens in production to trigger restarts by scaling up the size of each node, rather than adding new nodes to the cluster. E.g. scaling up to 4 nodes

<img width="1298" alt="image" src="https://user-images.githubusercontent.com/882976/220811254-6798acdd-761c-4b6b-aba3-4b4a0f0941f0.png">

And then back down to 2

<img width="1264" alt="image" src="https://user-images.githubusercontent.com/882976/220811266-dc9eb4b2-ade6-4d1c-827c-8fc28f199710.png">

This way the "scale up" and "scale down" will _both_ restart all nodes in the clusters. Should make testing a little more efficient, I hope!

**Comment by philip-stoev** [comment]
@pH14 Unfortunately your branch `pH14:persist-crdb-tcp-keepalives` did not seem to work as expected, a 2 min 11-second stall happened almost immediately as soon as I initiated a rolling CRDB restart:

```
CREATE TABLE
Time: 131243.539 ms (02:11.244)
```

CRDB was operational during that time in the sense that it accepted inserts into the consensus table.

**Comment by pH14** [comment]
@philip-stoev when you're up for it, I pushed another change to `pH14:persist-crdb-tcp-keepalives` that uses the `rust-postgres` branch and sets a tcp user timeout of 8s

**Comment by pH14** [comment]
I played with this more using Philip's repro and the `stall-test` Cockroach Cloud cluster via a scratch machine. This was very revealing to both the underlying problem, and to how I sent Philip on a wild goose chase with my PR in https://github.com/MaterializeInc/materialize/issues/17680#issuecomment-1441408555 which contained the wrong timeout values 🙈 

Using that branch + Philip's drop/create script + triggering a rolling restart on CRDB, I was able to quickly hit the 120+s DDL statements. In another session, mashing `ss -o -t` in a loop revealed sockets stuck in `SYN-SENT` state:

```
SYN-SENT  0  1   $ip:$port  $cockroachcloud:26257   timer:(on,1min3sec,6)
```

Here we can see a socket trying to establish itself to Cockroach Cloud, on its 6th retry of 63s. From https://blog.cloudflare.com/when-tcp-sockets-refuse-to-die/ we see that SYN-SENT defaults to 6 retries with backoffs of 1,3,7,15,31,63s. https://man7.org/linux/man-pages/man7/tcp.7.html says of SYN retries: `The default value is 6, which corresponds to retrying for up to approximately 127 seconds.`, the timing of which maps on very closely to Philip's initial findings.

As soon as this socket hit its SYN-SENT retry count, I near-simultaneously saw a DDL statement finish:

```
CREATE TABLE
Time: 121574.628 ms (02:01.575)
```

I tried again removing all client-side + keepalive timeouts except for the `tokio-postgres` connection timeout, set to 2s, and saw no DDL take more than 9s. Since the slowness seems to be SYN retries during connection establishment, TCP keep-alives and user timeouts won't apply, nor does it look like we need app-level timeouts (for this problem at least!).

---

tl;dr - all we need to do is set [connect_timeout](https://github.com/MaterializeInc/materialize/pull/17713/files#diff-a8a0c717785a75fec3c6bfb2f6d5e6635fce589ddd4a483955f47804bc39d0ebR184) to something small and our problems will go away (cc @mjibson)
